docs/docs/lang.md

```diff
@@ -470,6 +470,69 @@ The result:
 +-------+
 ```
 
+## <a name="comprehensions"></a> Comprehensions
+
+Comprehensions provide a concise way of building [Composite Values](#composite-values) from sub-queries.
+
+Like [Rules](#rules), comprehensions consist of a head and a body. The body of a comprehension can be understood in exactly the same way as the body of a rule, that is, one or more expressions that must all be true in order for the overall body to be true. When the body evaluates to true, the head of the comprehension is evaluated to produce an element in the result.
+
+The body of a comprehension is able to refer to variables defined in the outer body. For example:
+
+```
+> region = "west", names = [name | sites[i].region = region, sites[i].name = name]
++-----------------+--------+
+|      NAMES      | REGION |
++-----------------+--------+
+| ["smoke","dev"] | "west" |
++-----------------+--------+
+```
+
+In the above query, the second expression contains an [Array Comprehension](#array-comprehension) that refers to the "region" variable. The region variable will be bound in the outer body.
+
+> When a comprehension refers to a variable in an outer body, OPA will reorder expressions in the outer body so that variables referred to in the comprehension are bound by the time the comprehension is evaluated.
+
+Comprehensions are similar to the same constructs found in other languages like Python. For example, we could write the above comprehension in Python as follows:
+
+```python
+region = "west"
```

Comment:
Could we add a comment like "# Python equivalent of Rego comprehension shown above" so that a casual reader will not think she can write this expression in Rego.